CI is red on `ci / test` and `benchmark`, and neither looks related to this diff. Notes so a reviewer does not have to redo the digging:

`ci / test` failed twice with a different suite each time. First run: `test/integration/content.test.ts` (500s) plus `standalone-modules-prerender.test.ts` with `Server process exited before becoming ready`. Rerun: `test/integration/base.test.ts`. Both suites pass here on this branch, 7/7 and 6/6. Reads like fixture servers contending on the runner.

`benchmark` never gets a server up: `Cannot find module '../dist/babel.cjs'`, repeated for every request. That is an install that did not build a dependency, not a perf regression.

The diff is 106 added lines over two files, one new exported function that nothing in this repo calls yet, plus a docs page. `main` was green this morning on the same suites.

You have `pnpm-lock.yaml` and `pnpm-workspace.yaml` modified but uncommitted on `main` locally. If that is the dependency fix, it would explain the babel one.

--- a/packages/shared/build.config.ts
+++ b/packages/shared/build.config.ts
@@ -5,6 +5,7 @@ export default defineBuildConfig({
   failOnWarn: false,
   entries: [
     { input: 'src/content', name: 'content' },
+    { input: 'src/content-runtime', name: 'content-runtime' },
     { input: 'src/kit', name: 'kit' },
     { input: 'src/devtools', name: 'devtools' },
     { input: 'src/i18n', name: 'i18n' },
@@ -19,6 +20,7 @@ export default defineBuildConfig({
     /^h3/,
     /^@nuxtjs\/i18n/,
     /^@nuxt\/content/,
+    /^@harlan-zw\/comark-content/,
     /^@nuxt\/schema/,
     /^@nuxt\/devtools-kit/,
     /^@nuxt\/kit/,

--- a/packages/shared/package.json
+++ b/packages/shared/package.json
@@ -28,6 +28,10 @@
       "types": "./dist/content.d.mts",
       "default": "./dist/content.mjs"
     },
+    "./content-runtime": {
+      "types": "./dist/content-runtime.d.mts",
+      "default": "./dist/content-runtime.mjs"
+    },
     "./kit": {
       "types": "./dist/kit.d.mts",
       "default": "./dist/kit.mjs"
@@ -110,6 +114,8 @@
   },
   "devDependencies": {
     "@arethetypeswrong/cli": "catalog:",
+    "@harlan-zw/comark-content": "catalog:",
+    "@nuxt/content": "catalog:",
     "@nuxt/module-builder": "catalog:",
     "@nuxt/schema": "catalog:",
     "@nuxt/test-utils": "catalog:",

--- a/packages/shared/src/content-runtime.ts
+++ b/packages/shared/src/content-runtime.ts
@@ -1,0 +1,45 @@
+/**
+ * Contract for `#nuxtseo/content`, the per-provider runtime shim that
+ * `setupContentRuntime()` aliases into the Nitro bundle.
+ *
+ * `@nuxt/content` and `comark-content` cannot both be imported by one build:
+ * pulling `@nuxt/content/server` into an app that does not have it fails the
+ * bundle. So provider choice is a build-time alias, and every consumer imports
+ * this one specifier instead of naming a package.
+ */
+
+export type ContentRuntimeProvider = 'nuxt-content-v3' | 'comark' | 'none'
+
+/**
+ * A queryable page collection, normalized across providers.
+ *
+ * The two manifests disagree in shape. `@nuxt/content` keys an object by name and
+ * carries a schema-derived `fields` map. comark returns an array and carries a
+ * per-collection sitemap flag. Both reduce to this.
+ */
+export interface PageCollection {
+  name: string
+  /**
+   * Whether the collection declares `field`.
+   *
+   * For `@nuxt/content` this is the schema opt in, so a module can require its own
+   * field before claiming the collection. comark derives no field list from a
+   * schema, so it answers `true` and relies on frontmatter being present or absent.
+   */
+  hasField: (field: string) => boolean
+  /** `false` only when the collection opted out of the sitemap. */
+  inSitemap: boolean
+}
+
+export interface ContentPageQuery {
+  where: (field: string, operator: string, value?: unknown) => ContentPageQuery
+  select: (...fields: string[]) => ContentPageQuery
+  all: () => Promise<Record<string, any>[]>
+  first: () => Promise<Record<string, any> | null>
+}
+
+export interface ContentRuntime {
+  provider: ContentRuntimeProvider
+  listPageCollections: (event: unknown) => Promise<PageCollection[]>
+  queryPages: (event: unknown, collection: string) => ContentPageQuery
+}

--- a/packages/shared/src/kit.ts
+++ b/packages/shared/src/kit.ts
@@ -314,3 +314,35 @@ export async function resolveContentProvider(nuxt: Nuxt = useNuxt()): Promise<Co
 export function hasContentFileHooks(provider: ContentProvider): boolean {
   return provider._tag === 'Comark' || (provider._tag === 'NuxtContent' && provider.version === 3)
 }
+
+/**
+ * Alias `#nuxtseo/content` to the shim for the detected provider.
+ *
+ * Consumers import collection enumeration and page queries from that one
+ * specifier instead of naming `@nuxt/content/server` or
+ * `@harlan-zw/comark-content/server`. Naming a package directly would put it in
+ * every build, and a build without that package installed fails to bundle.
+ *
+ * Nuxt Content v2 has no collection model, so it resolves to the empty shim.
+ * A module that supports v2 keeps its own v2 path.
+ */
+export function setupContentRuntime(provider: ContentProvider, nuxt: Nuxt = useNuxt()): void {
+  const shim = provider._tag === 'Comark'
+    ? 'comark'
+    : provider._tag === 'NuxtContent' && provider.version === 3
+      ? 'nuxt-content-v3'
+      : 'none'
+  const resolver = createResolver(import.meta.url)
+  // This package does not depend on nitro's option types, so narrow the two fields
+  // it touches rather than pulling the whole NitroConfig augmentation in.
+  const nitro = (nuxt.options as { nitro?: { alias?: Record<string, string>, externals?: { inline?: string[] } } }).nitro ??= {}
+  nitro.alias ??= {}
+  nitro.alias['#nuxtseo/content'] = resolver.resolve(`./runtime/content/${shim}`)
+  // The shim ships inside this package, so Nitro treats it as an external dependency
+  // and leaves its `@nuxt/content/server` import for Node to resolve at runtime.
+  // `#content/manifest` is a build-time alias, so that import then fails to load.
+  // Inline the shim directory to get it bundled with the aliases applied.
+  nitro.externals ??= {}
+  nitro.externals.inline ??= []
+  nitro.externals.inline.push(resolver.resolve('./runtime/content/'))
+}

--- a/packages/shared/src/runtime/content/comark.ts
+++ b/packages/shared/src/runtime/content/comark.ts
@@ -1,0 +1,55 @@
+import type { ContentPageQuery, ContentRuntimeProvider, PageCollection } from '../../content-runtime'
+import { queryCollection, queryCollectionManifest } from '@harlan-zw/comark-content/server'
+
+export const provider: ContentRuntimeProvider = 'comark'
+
+export async function listPageCollections(event: unknown): Promise<PageCollection[]> {
+  const manifest = await queryCollectionManifest(event)
+  return manifest.map(entry => ({
+    name: entry.name,
+    // comark derives no field list from a schema, so every field is available to ask
+    // for. A collection opts out of the sitemap on the collection itself.
+    hasField: () => true,
+    inSitemap: entry.sitemap,
+  }))
+}
+
+/**
+ * comark's query builder gained `IS NOT NULL` in 0.1.4. Translate it into a
+ * post-filter so one caller works against every supported comark.
+ *
+ * Every operation is buffered rather than forwarded on call, so the caller can
+ * order `where` and `select` however it likes: a post-filtered field has to survive
+ * the projection to be readable, and that is only known once the chain ends.
+ */
+export function queryPages(event: unknown, collection: string): ContentPageQuery {
+  const wheres: [string, string, unknown][] = []
+  let selected: string[] | undefined
+
+  const query: ContentPageQuery = {
+    where(field, operator, value) {
+      wheres.push([field, operator, value])
+      return query
+    },
+    select(...fields) {
+      selected = fields
+      return query
+    },
+    async all() {
+      const notNull = wheres.filter(([, operator]) => operator === 'IS NOT NULL').map(([field]) => field)
+      const builder = queryCollection(event, collection)
+      for (const [field, operator, value] of wheres) {
+        if (operator !== 'IS NOT NULL')
+          builder.where(field, operator as Parameters<typeof builder.where>[1], value)
+      }
+      if (selected)
+        builder.select(...new Set([...selected, ...notNull]))
+      const rows = await builder.all() as Record<string, any>[]
+      return rows.filter(row => notNull.every(field => row[field] !== null && row[field] !== undefined))
+    },
+    async first() {
+      return (await query.all())[0] ?? null
+    },
+  }
+  return query
+}

--- a/packages/shared/src/runtime/content/none.ts
+++ b/packages/shared/src/runtime/content/none.ts
@@ -1,0 +1,15 @@
+import type { ContentPageQuery, ContentRuntimeProvider, PageCollection } from '../../content-runtime'
+
+export const provider: ContentRuntimeProvider = 'none'
+
+export async function listPageCollections(_event: unknown): Promise<PageCollection[]> {
+  return []
+}
+
+/**
+ * Never reached: every caller enumerates collections first and gets none.
+ * Present so the specifier resolves and consumers import it unconditionally.
+ */
+export function queryPages(_event: unknown, collection: string): ContentPageQuery {
+  throw new Error(`No content module is installed, so collection "${collection}" cannot be queried.`)
+}

--- a/packages/shared/src/runtime/content/nuxt-content-v3.ts
+++ b/packages/shared/src/runtime/content/nuxt-content-v3.ts
@@ -1,0 +1,21 @@
+import type { ContentPageQuery, ContentRuntimeProvider, PageCollection } from '../../content-runtime'
+import { queryCollection } from '@nuxt/content/server'
+// @ts-expect-error virtual module provided by @nuxt/content
+import manifest from '#content/manifest'
+
+export const provider: ContentRuntimeProvider = 'nuxt-content-v3'
+
+export async function listPageCollections(_event: unknown): Promise<PageCollection[]> {
+  return Object.entries(manifest as Record<string, { fields?: Record<string, unknown> }>)
+    .map(([name, entry]) => ({
+      name,
+      hasField: (field: string) => !!entry.fields && field in entry.fields,
+      // @nuxt/content has no collection level sitemap opt out. The schema field is
+      // the opt in, which `hasField` answers.
+      inSitemap: true,
+    }))
+}
+
+export function queryPages(event: unknown, collection: string): ContentPageQuery {
+  return queryCollection(event as never, collection as never) as unknown as ContentPageQuery
+}

--- a/pnpm-workspace.yaml
+++ b/pnpm-workspace.yaml
@@ -61,6 +61,7 @@ minimumReleaseAgeExclude:
   - '@unhead/vue@3.2.1'
   - '@mdream/rust-wasm32-wasi@1.5.12'
   - eslint-plugin-harlanzw@0.19.1
+  - '@harlan-zw/comark-content@0.1.3'
 # pnpm 11 reads these from pnpm-workspace.yaml, not .npmrc.
 # @nuxtjs/i18n resolves its private @intlify/* + vue-i18n deps via @nuxt/kit's
 # resolveModule, which only searches root node_modules, so hoist just those
@@ -95,6 +96,7 @@ catalog:
   '@clack/prompts': ^1.7.0
   '@iconify-json/carbon': ^1.2.25
   '@jridgewell/trace-mapping': ^0.3.31
+  '@harlan-zw/comark-content': ^0.1.3
   '@nuxt/content': ^3.15.2
   '@nuxt/devtools-kit': 4.0.0-alpha.7
   '@nuxt/fonts': ^0.14.0

--- a/pnpm-workspace.yaml
+++ b/pnpm-workspace.yaml
@@ -28,7 +28,7 @@ minimumReleaseAgeExclude:
   - '@img/sharp-win32-arm64@0.35.0'
   - '@img/sharp-win32-ia32@0.35.0'
   - '@img/sharp-win32-x64@0.35.0'
-  - '@nuxtjs/robots@6.1.0 || 6.1.3'
+  - '@nuxtjs/robots@6.1.0 || 6.1.3 || 6.1.5'
   - '@nuxtjs/sitemap@8.2.0 || 8.2.3 || 8.3.0'
   - nuxt-ai-ready@1.4.0
   - nuxt-link-checker@5.1.0 || 5.1.3
@@ -94,9 +94,9 @@ catalog:
   '@antfu/eslint-config': ^9.3.0
   '@arethetypeswrong/cli': ^0.18.5
   '@clack/prompts': ^1.7.0
+  '@harlan-zw/comark-content': ^0.1.3
   '@iconify-json/carbon': ^1.2.25
   '@jridgewell/trace-mapping': ^0.3.31
-  '@harlan-zw/comark-content': ^0.1.3
   '@nuxt/content': ^3.15.2
   '@nuxt/devtools-kit': 4.0.0-alpha.7
   '@nuxt/fonts': ^0.14.0
@@ -106,7 +106,7 @@ catalog:
   '@nuxt/test-utils': ^4.1.0
   '@nuxt/ui': ^4.10.0
   '@nuxtjs/i18n': ^10.6.0
-  '@nuxtjs/robots': ^6.1.4
+  '@nuxtjs/robots': ^6.1.5
   '@nuxtjs/sitemap': ^8.3.3
   '@resvg/resvg-js': ^2.6.2
   '@shikijs/langs': ^4.4.2

--- a/pnpm-workspace.yaml
+++ b/pnpm-workspace.yaml
@@ -28,7 +28,7 @@ minimumReleaseAgeExclude:
   - '@img/sharp-win32-arm64@0.35.0'
   - '@img/sharp-win32-ia32@0.35.0'
   - '@img/sharp-win32-x64@0.35.0'
-  - '@nuxtjs/robots@6.1.0 || 6.1.3 || 6.1.5'
+  - '@nuxtjs/robots@6.1.0 || 6.1.3'
   - '@nuxtjs/sitemap@8.2.0 || 8.2.3 || 8.3.0'
   - nuxt-ai-ready@1.4.0
   - nuxt-link-checker@5.1.0 || 5.1.3
@@ -106,7 +106,7 @@ catalog:
   '@nuxt/test-utils': ^4.1.0
   '@nuxt/ui': ^4.10.0
   '@nuxtjs/i18n': ^10.6.0
-  '@nuxtjs/robots': ^6.1.5
+  '@nuxtjs/robots': ^6.1.4
   '@nuxtjs/sitemap': ^8.3.3
   '@resvg/resvg-js': ^2.6.2
   '@shikijs/langs': ^4.4.2

